File: src/Shared/ECSPros.Shared.Kernel/Common/I18nHelper.cs
namespace ECSPros.Shared.Kernel.Common;

/// <summary>
/// Çok dilli sözlüklerden dil + cinsiyet bağlamına göre değer döndüren yardımcı.
/// Anahtar formatı: "{lang}" veya "{lang}:{cinsiyet}" (örn. "ar:m", "ar:f")
/// </summary>
public static class I18nHelper
{
    /// <summary>
    /// Cinsiyet varyantı destekleyen dil kodları.
    /// </summary>
    public static readonly IReadOnlySet<string> GenderAwareLangs =
        new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "ar", "he" };

    /// <summary>
    /// Sözlükten en uygun değeri döndürür.
    /// Öncelik: lang:genderCtx → lang → fallbackLang → ilk değer
    /// </summary>
    /// <param name="i18n">Flat sözlük: {"tr":"Pantolon","ar":"بنطلون","ar:f":"بنطال"}</param>
    /// <param name="lang">İstenen dil kodu (örn. "ar")</param>
    /// <param name="genderCtx">"m" | "f" | null</param>
    /// <param name="fallbackLang">Yedek dil (varsayılan: "tr")</param>
    public static string GetLocalized(
        Dictionary<string, string>? i18n,
        string lang,
        string? genderCtx = null,
        string fallbackLang = "tr")
    {
        if (i18n is null || i18n.Count == 0) return string.Empty;

        // 1. lang:gender → "ar:f"
        if (!string.IsNullOrEmpty(genderCtx) &&
            i18n.TryGetValue($"{lang}:{genderCtx}", out var genVal))
            return genVal;

        // 2. lang → "ar"
        if (i18n.TryGetValue(lang, out var langVal))
            return langVal;

        // 3. fallback dil
        if (i18n.TryGetValue(fallbackLang, out var fbVal))
            return fbVal;

        // 4. ilk değer
        return i18n.Values.FirstOrDefault() ?? string.Empty;
    }
}

File: admin/src/lib/i18n-helper.ts
/**
 * i18n-helper — Cinsiyet bağlamına duyarlı çok dilli yardımcı fonksiyonlar.
 *
 * Anahtar formatı: "{lang}" veya "{lang}:{cinsiyet}"
 *   Örnek: { "tr": "Pantolon", "ar": "بنطلون", "ar:m": "بنطلون رجالي", "ar:f": "بنطلون نسائي" }
 */

/** Cinsiyet varyantı destekleyen dil kodları */
export const GENDER_AWARE_LANGS: ReadonlySet<string> = new Set(['ar', 'he'])

/**
 * Çok dilli sözlükten en uygun değeri döndürür.
 * Öncelik: lang:genderCtx → lang → fallback → ilk değer
 */
export function getLocalized(
  i18n: Record<string, string> | undefined,
  lang: string,
  genderCtx?: 'm' | 'f' | null,
  fallback = 'tr',
): string {
  if (!i18n) return ''
  if (genderCtx && i18n[`${lang}:${genderCtx}`]) return i18n[`${lang}:${genderCtx}`]
  if (i18n[lang]) return i18n[lang]
  if (i18n[fallback]) return i18n[fallback]
  return Object.values(i18n)[0] ?? ''
}

/**
 * I18nField bileşeni için değer haritası oluşturur.
 * Cinsiyet farkında diller için "lang:m" ve "lang:f" pseudo-anahtarlarını da içerir.
 *
 * @param nameI18n  Flat dict: { tr: "Pantolon", ar: "بنطلون", "ar:m": "...", "ar:f": "..." }
 * @param languages Dil listesi (genderAware flag içerir)
 * @param fieldKey  Tek alan anahtarı (varsayılan: "name")
 */
export function buildI18nValues(
  nameI18n: Record<string, string>,
  languages: ReadonlyArray<{ code: string; genderAware?: boolean }>,
  fieldKey = 'name',
): Record<string, Record<string, string>> {
  const result: Record<string, Record<string, string>> = {}
  for (const l of languages) {
    result[l.code] = { [fieldKey]: nameI18n[l.code] ?? '' }
    if (l.genderAware) {
      result[`${l.code}:m`] = { [fieldKey]: nameI18n[`${l.code}:m`] ?? '' }
      result[`${l.code}:f`] = { [fieldKey]: nameI18n[`${l.code}:f`] ?? '' }
    }
  }
  return result
}
